client/state/happychat/chat/reducer.js

```diff
@@ -88,66 +83,70 @@ const timelineEvent = ( state = {}, action ) => {
 				files: message.files,
 				timestamp: maybeUpscaleTimePrecision( message.timestamp ),
 				user_id: message.user.id,
-				type: get( message, 'type', 'message' ),
-				links: get( message, 'meta.links' ),
+				type: message.type ?? 'message',
+				links: message.meta?.links,
 			};
+		}
 	}
 	return state;
 };
 
-const sortTimeline = ( timeline ) =>
-	sortBy( timeline, ( event ) => parseInt( event.timestamp, 10 ) );
+const getEventTimestamp = ( event ) => parseInt( event.timestamp, 10 );
+const timelineSortCmp = ( a, b ) => getEventTimestamp( a ) - getEventTimestamp( b );
+const sortTimeline = ( timeline ) => timeline.concat().sort( timelineSortCmp );
 
 /**
  * Adds timeline events for happychat
  *
  * @param  {object} state  Current state
  * @param  {object} action Action payload
  * @returns {object}        Updated state
- *
  */
 const timelineReducer = ( state = [], action ) => {
 	switch ( action.type ) {
 		case HAPPYCHAT_IO_RECEIVE_MESSAGE:
-		case HAPPYCHAT_IO_RECEIVE_MESSAGE_OPTIMISTIC:
+		case HAPPYCHAT_IO_RECEIVE_MESSAGE_OPTIMISTIC: {
 			// if meta.forOperator is set, skip so won't show to user
-			if ( get( action, 'message.meta.forOperator', false ) ) {
+			if ( action.message.meta?.forOperator ?? false ) {
 				return state;
 			}
 			const event = timelineEvent( {}, action );
 
 			// If the message already exists in the timeline, replace it
-			const idx = findIndex( state, ( { id } ) => event.id === id );
+			const idx = state.findIndex( ( { id } ) => event.id === id );
 			if ( idx >= 0 ) {
 				return [ ...state.slice( 0, idx ), event, ...state.slice( idx + 1 ) ];
 			}
 
 			// This is a new message — append it!
-			return concat( state, [ event ] );
-
-		case HAPPYCHAT_IO_RECEIVE_MESSAGE_UPDATE:
-			const index = findIndex( state, ( { id } ) => action.message.id === id );
+			return state.concat( [ event ] );
+		}
+		case HAPPYCHAT_IO_RECEIVE_MESSAGE_UPDATE: {
+			const index = state.findIndex( ( { id } ) => action.message.id === id );
 			return index === -1
 				? state
 				: [ ...state.slice( 0, index ), timelineEvent( {}, action ), ...state.slice( index + 1 ) ];
-		case HAPPYCHAT_IO_REQUEST_TRANSCRIPT_TIMEOUT:
+		}
+		case HAPPYCHAT_IO_REQUEST_TRANSCRIPT_TIMEOUT: {
 			return state;
-		case HAPPYCHAT_IO_REQUEST_TRANSCRIPT_RECEIVE:
-			const messages = filter( action.messages, ( message ) => {
-				if ( ! message.id ) {
-					return false;
-				}
+		}
+		case HAPPYCHAT_IO_REQUEST_TRANSCRIPT_RECEIVE: {
+			const messages =
+				action.messages?.filter( ( message ) => {
```

Comment:
If there are no `action.messages`, you could `return state` at the very beginning and then proceed assuming `action.messages` is valid.

Reply:
What if `state` contains old messages and `action.messages` is empty? Shouldn't we return `[]` in that case to clear the old messages?

Reply:
I think I'll keep it as-is for the time being. The optional chaining is just a precaution because the messages come from the API and they could be empty for some reason.

Reply:
> What if state contains old messages and action.messages is empty? Shouldn't we return [] in that case to clear the old messages?

That would be true if the "receive transcript" action replaced the existing `state`, but it merely appends messages that aren't there yet and sorts them by time. Then empty `action.messages` means no update.